File: lib/tools/analytics.ts
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { getSupabase } from "../supabase";
import {
  getSessionFrequencyInput,
  getVolumeSummaryInput,
} from "../schema";
import { ok, fail } from "./shared";

type SetRow = {
  exercise_name: string;
  weight_kg: number | null;
  reps: number | null;
  session: { date: string } | { date: string }[] | null;
};

function mondayOf(date: Date): Date {
  const d = new Date(Date.UTC(
    date.getUTCFullYear(),
    date.getUTCMonth(),
    date.getUTCDate()
  ));
  const day = d.getUTCDay(); // 0 Sun .. 6 Sat
  const diff = (day === 0 ? -6 : 1 - day);
  d.setUTCDate(d.getUTCDate() + diff);
  return d;
}

export function registerAnalyticsTools(server: McpServer) {
  server.tool(
    "get_volume_summary",
    "Total sets, reps, and tonnage between two dates, grouped by exercise.",
    getVolumeSummaryInput,
    async (input) => {
      try {
        const supabase = getSupabase();
        let query = supabase
          .from("exercise_sets")
          .select(
            "exercise_name, weight_kg, reps, session:workout_sessions!inner(date)"
          )
          .gte("session.date", input.from_date)
          .lte("session.date", input.to_date);

        if (input.exercise_name) {
          query = query.eq("exercise_name", input.exercise_name);
        }

        const { data, error } = await query;
        if (error) throw error;

        type Bucket = {
          exercise_name: string;
          total_sets: number;
          total_reps: number;
          total_tonnage_kg: number | null;
          has_bodyweight_set: boolean;
        };

        const buckets = new Map<string, Bucket>();
        for (const row of (data ?? []) as SetRow[]) {
          const key = row.exercise_name;
          let b = buckets.get(key);
          if (!b) {
            b = {
              exercise_name: key,
              total_sets: 0,
              total_reps: 0,
              total_tonnage_kg: 0,
              has_bodyweight_set: false,
            };
            buckets.set(key, b);
          }

          b.total_sets += 1;
          b.total_reps += row.reps ?? 0;

          if (row.weight_kg === null || row.weight_kg === undefined) {
            b.has_bodyweight_set = true;
          } else if (b.total_tonnage_kg !== null && row.reps !== null) {
            b.total_tonnage_kg += Number(row.weight_kg) * Number(row.reps);
          }
        }

        // If any set in a bucket was bodyweight, tonnage becomes null.
        const summary = [...buckets.values()].map((b) => ({
          exercise_name: b.exercise_name,
          total_sets: b.total_sets,
          total_reps: b.total_reps,
          total_tonnage_kg: b.has_bodyweight_set ? null : b.total_tonnage_kg,
        }));

        return ok({
          from_date: input.from_date,
          to_date: input.to_date,
          summary,
        });
      } catch (err) {
        console.error("get_volume_summary failed", { input, err });
        return fail("get_volume_summary", err);
      }
    }
  );

  server.tool(
    "get_session_frequency",
    "Sessions per ISO week for the last N weeks (default 8).",
    getSessionFrequencyInput,
    async (input) => {
      try {
        const weeks = input.weeks ?? 8;
        const supabase = getSupabase();

        const now = new Date();
        const startMonday = mondayOf(now);
        startMonday.setUTCDate(startMonday.getUTCDate() - 7 * (weeks - 1));

        const { data, error } = await supabase
          .from("workout_sessions")
          .select("date")
          .gte("date", startMonday.toISOString())
          .order("date", { ascending: true });

        if (error) throw error;

        const counts = new Map<string, number>();
        for (let i = 0; i < weeks; i++) {
          const w = new Date(startMonday);
          w.setUTCDate(startMonday.getUTCDate() + 7 * i);
          counts.set(w.toISOString().slice(0, 10), 0);
        }

        for (const row of data ?? []) {
          const w = mondayOf(new Date(row.date)).toISOString().slice(0, 10);
          if (counts.has(w)) counts.set(w, (counts.get(w) ?? 0) + 1);
        }

        const result = [...counts.entries()]
          .sort(([a], [b]) => a.localeCompare(b))
          .map(([week_start, count]) => ({ week_start, count }));

        return ok({ weeks, frequency: result });
      } catch (err) {
        console.error("get_session_frequency failed", { input, err });
        return fail("get_session_frequency", err);
      }
    }
  );
}

File: apps/api/app/services/analytics.py
"""Analytics: training volume by exercise and session frequency by ISO week.

Volume tonnage sums ``weight_kg × reps`` per exercise, but is reported as ``None`` when
any set in the range was bodyweight (null weight) — a partial tonnage would mislead.
Frequency counts sessions into Monday-anchored weekly buckets for the last N weeks.
"""

from __future__ import annotations

import uuid
from collections.abc import Sequence
from dataclasses import dataclass
from datetime import UTC, date, datetime, timedelta

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core import errors
from app.models import Exercise, ExerciseSet, WorkoutSession


@dataclass(frozen=True)
class VolumeBucket:
    exercise_id: uuid.UUID
    exercise_name: str
    total_sets: int
    total_reps: int
    total_tonnage_kg: float | None


@dataclass(frozen=True)
class WeekCount:
    week_start: date
    count: int


async def volume(
    db: AsyncSession,
    *,
    user_id: uuid.UUID,
    date_from: datetime,
    date_to: datetime,
    exercise_id: uuid.UUID | None = None,
) -> Sequence[VolumeBucket]:
    """Total sets, reps, and tonnage per exercise between two instants (inclusive).

    Aggregated in SQL. It used to select one row per *set* and fold them in Python, so a year's
    range shipped thousands of rows across the wire to produce a few dozen — and the cost grew
    with training history rather than with the answer. Now Postgres returns one row per exercise.

    Two details in the SQL carry the original semantics exactly:

    * ``coalesce`` around each ``sum`` — an empty group would otherwise be ``NULL``, and ``None``
      already means something specific in this API (see below).
    * ``bool_or(weight_kg IS NULL)`` — tonnage is reported as ``None``, not a smaller number, when
      *any* set in the group was bodyweight. A partial tonnage looks like a real total and would
      quietly understate the work done.
    """
    if date_from > date_to:
        raise errors.validation("date_from must be on or before date_to")

    stmt = (
        select(
            ExerciseSet.exercise_id,
            Exercise.name,
            func.count().label("total_sets"),
            func.coalesce(func.sum(func.coalesce(ExerciseSet.reps, 0)), 0).label("total_reps"),
            # NULL × anything is NULL, so a bodyweight set or one with no reps contributes 0 —
            # the same two guards the Python version spelled out as branches.
            func.coalesce(
                func.sum(func.coalesce(ExerciseSet.weight_kg * ExerciseSet.reps, 0)), 0
            ).label("tonnage"),
            func.bool_or(ExerciseSet.weight_kg.is_(None)).label("has_bodyweight"),
        )
        .join(WorkoutSession, ExerciseSet.session_id == WorkoutSession.id)
        .join(Exercise, ExerciseSet.exercise_id == Exercise.id)
        .where(
            ExerciseSet.user_id == user_id,
            WorkoutSession.performed_at >= date_from,
            WorkoutSession.performed_at <= date_to,
        )
        .group_by(ExerciseSet.exercise_id, Exercise.name)
    )
    if exercise_id is not None:
        stmt = stmt.where(ExerciseSet.exercise_id == exercise_id)

    result = [
        VolumeBucket(
            exercise_id=ex_id,
            exercise_name=name,
            total_sets=total_sets,
            total_reps=total_reps,
            total_tonnage_kg=None if has_bodyweight else round(float(tonnage), 3),
        )
        for ex_id, name, total_sets, total_reps, tonnage, has_bodyweight in (
            await db.execute(stmt)
        ).all()
    ]
    # Sorted in Python, deliberately: the database orders by its collation, which disagrees with
    # Python's codepoint order on case and punctuation. Letting it sort would silently change the
    # sequence MCP clients already see.
    result.sort(key=lambda b: b.exercise_name)
    return result


def _monday_of(moment: datetime) -> date:
    """The Monday (UTC date) of the week containing ``moment``."""
    day = moment.astimezone(UTC).date()
    return day - timedelta(days=day.weekday())


async def frequency(
    db: AsyncSession,
    *,
    user_id: uuid.UUID,
    weeks: int = 8,
    now: datetime | None = None,
) -> Sequence[WeekCount]:
    """Sessions per ISO week for the last ``weeks`` weeks (Monday-anchored, zero-filled)."""
    if not 1 <= weeks <= 52:
        raise errors.validation("weeks must be between 1 and 52")

    current = now or datetime.now(tz=UTC)
    start_monday = _monday_of(current) - timedelta(weeks=weeks - 1)

    counts: dict[date, int] = {start_monday + timedelta(weeks=i): 0 for i in range(weeks)}

    performed_ats = (
        (
            await db.execute(
                select(WorkoutSession.performed_at).where(
                    WorkoutSession.user_id == user_id,
                    WorkoutSession.performed_at
                    >= datetime.combine(start_monday, datetime.min.time(), tzinfo=UTC),
                )
            )
        )
        .scalars()
        .all()
    )
    for performed_at in performed_ats:
        bucket = _monday_of(performed_at)
        if bucket in counts:
            counts[bucket] += 1

    return [WeekCount(week_start=week, count=counts[week]) for week in sorted(counts)]
